File: services/voice-gateway/src/speak.ts
import fs from 'node:fs'
import path from 'node:path'
import { fileURLToPath } from 'node:url'
import type { SpeakStyle, SpeakTarget, VoiceChannel } from '@forkai/shared'
import type { GatewayConfig } from './config.js'

const __dirname = path.dirname(fileURLToPath(import.meta.url))

export class SpeakService {
  private templates: Record<string, string>

  constructor(private cfg: GatewayConfig) {
    const p = path.resolve(__dirname, '../config/utterances.zh-CN.json')
    this.templates = JSON.parse(fs.readFileSync(p, 'utf8'))
  }

  render(key: string, params: Record<string, string | number> = {}): string {
    let text = this.templates[key] || key
    for (const [k, v] of Object.entries(params)) {
      text = text.replace(new RegExp(`\\{${k}\\}`, 'g'), String(v))
    }
    return text
  }

  resolveTarget(
    channel: VoiceChannel,
    kind: 'move' | 'query' | 'wake' | 'fail'
  ): SpeakTarget {
    if (kind === 'wake') return 'vehicle'
    if (kind === 'query') return this.cfg.speak.query
    if (channel === 'cabin') return this.cfg.speak.cabinMove
    return this.cfg.speak.pttMove
  }

  async speak(opts: {
    text: string
    style: SpeakStyle
    target: SpeakTarget
    clientId?: string
  }): Promise<{ text: string; audioBase64?: string }> {
    try {
      const res = await fetch(`${this.cfg.speech.baseUrl.replace(/\/$/, '')}/tts/speak`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({
          text: opts.text,
          style: opts.style,
          target: opts.target
        })
      })
      if (!res.ok) return { text: opts.text }
      const data = (await res.json()) as { audioBase64?: string }
      return { text: opts.text, audioBase64: data.audioBase64 }
    } catch {
      return { text: opts.text }
    }
  }
}

File: services/core/app/speak.py
"""话术渲染与播报目标路由（对齐 speak.ts 的 render / resolveTarget）。

render：读 config/utterances.zh-CN.json，{n}/{goal} 等占位符逐参数 str.replace
（缺失参数不炸：模板中未提供的占位符原样保留，key 不存在时返回 key 本身）。
resolveTarget：wake→vehicle；query→speak.query；cabin→speak.cabinMove；ptt→speak.pttMove。
"""
import json

from .config import CORE_ROOT

_TEMPLATES_PATH = CORE_ROOT / "config" / "utterances.zh-CN.json"

with open(_TEMPLATES_PATH, "r", encoding="utf-8") as _f:
    TEMPLATES: dict = json.load(_f)


def render(key: str, params: dict | None = None) -> str:
    text = TEMPLATES.get(key, key)
    for k, v in (params or {}).items():
        text = text.replace("{" + str(k) + "}", str(v))
    return text


def resolve_target(channel: str, speak_kind: str, cfg: dict) -> str:
    if speak_kind == "wake":
        return "vehicle"
    if speak_kind == "query":
        return cfg["speak"]["query"]
    if channel == "cabin":
        return cfg["speak"]["cabinMove"]
    return cfg["speak"]["pttMove"]
